protocol/contracts/zaps/GenericZap.sol

```diff
@@ -0,0 +1,697 @@
+pragma solidity ^0.8.4;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+import "./ZapBase.sol";
+import "./libs/Swap.sol";
+//import "./libs/Executable.sol";
+import "./interfaces/IBalancerVault.sol";
+import "./interfaces/IUniswapV2Factory.sol";
+import "./interfaces/IUniswapV2Router.sol";
+import "./interfaces/ICurvePool.sol";
+import "./interfaces/ICurveFactory.sol";
+
+
+contract GenericZap is ZapBase {
+
+  IUniswapV2Router public immutable uniswapV2Router;
+
+  address private constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+  ICurveFactory private immutable curveFactory = ICurveFactory(0xB9fC157394Af804a3578134A6585C0dc9cc990d4);
+  IBalancerVault private immutable balancerVault = IBalancerVault(0xBA12222222228d8Ba445958a75a0704d566BF2C8);
+
+  struct ZapLiquidityRequest {
+    uint248 firstSwapMinAmountOut;
+    bool useAltFunction;
+    uint248 poolSwapMinAmountOut;
+    bool isOneSidedLiquidityAddition;
+    address otherToken;
+    bool shouldTransferResidual;
+    uint256 minLiquidityOut;
+    uint256 uniAmountAMin;
+    uint256 uniAmountBMin;
+    bytes poolSwapData;
+  }
+
+  event ZappedIn(address indexed sender, address fromToken, uint256 fromAmount, address toToken, uint256 amountOut);
+  event ZappedLPUniV2(address indexed recipient, address token0, address token1, uint256 amountA, uint256 amountB);
+  event TokenRecovered(address token, address to, uint256 amount);
+  event ZappedLPCurve(address indexed recipient, address fromToken, uint256 liquidity, uint256[] amounts);
+  event ZappedLiquidityBalancerPool(address indexed recipient, address fromToken, uint256 fromAmount, uint256[] maxAmountsIn);
+
+  constructor (
+    address _router
+  ) {
+    uniswapV2Router = IUniswapV2Router(_router);
+  }
+
+  /**
+   * @notice recover token or ETH
+   * @param _token token to recover
+   * @param _to receiver of recovered token
+   * @param _amount amount to recover
+   */
+  function recoverToken(address _token, address _to, uint256 _amount) external onlyOwner {
+    require(_to != address(0), "Invalid receiver");
+    if (_token == address(0)) {
+      // this is effectively how OpenZeppelin transfers eth
+      require(address(this).balance >= _amount, "Address: insufficient balance");
+      (bool success,) = _to.call{value: _amount}(""); 
```

Comment:
Hmm we can't recover ETH from stax contracts...worth us considering adding this to the Transfer lib?

Reply:
yes, it's worth adding that